File: scripts/enc.py
"""
Shared AES-256-GCM encryption for stored secrets (e.g. users' Gmail app
passwords). Key = SHA-256(AUTH_SECRET). Format: base64(iv[12] + ciphertext).
Must stay byte-compatible with webapp/functions/_shared/enc.js.
"""

from __future__ import annotations

import envload  # noqa: F401
import base64
import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

_SECRET = os.getenv("AUTH_SECRET", "").strip()


def _key() -> bytes:
    if not _SECRET:
        raise RuntimeError("AUTH_SECRET not set")
    return hashlib.sha256(_SECRET.encode()).digest()


def encrypt(plain: str) -> str:
    iv = os.urandom(12)
    ct = AESGCM(_key()).encrypt(iv, plain.encode(), None)
    return base64.b64encode(iv + ct).decode()


def decrypt(token: str) -> str:
    raw = base64.b64decode(token)
    return AESGCM(_key()).decrypt(raw[:12], raw[12:], None).decode()

File: webapp/functions/_shared/enc.js
// AES-256-GCM for stored secrets. Key = SHA-256(AUTH_SECRET).
// Format: base64(iv[12] + ciphertext). Byte-compatible with scripts/enc.py.

async function key(secret) {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(secret));
  return crypto.subtle.importKey("raw", digest, "AES-GCM", false, ["encrypt", "decrypt"]);
}

export async function encrypt(plain, secret) {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const k = await key(secret);
  const ct = new Uint8Array(await crypto.subtle.encrypt({ name: "AES-GCM", iv }, k, new TextEncoder().encode(plain)));
  const out = new Uint8Array(iv.length + ct.length);
  out.set(iv); out.set(ct, iv.length);
  let bin = ""; for (const b of out) bin += String.fromCharCode(b);
  return btoa(bin);
}

export async function decrypt(token, secret) {
  const bin = atob(token);
  const raw = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) raw[i] = bin.charCodeAt(i);
  const k = await key(secret);
  const pt = await crypto.subtle.decrypt({ name: "AES-GCM", iv: raw.slice(0, 12) }, k, raw.slice(12));
  return new TextDecoder().decode(pt);
}
